runtime/src/gpu/cpu/gpu-cpu.c

```diff
@@ -0,0 +1,158 @@
+  CHPL_GPU_DEBUG("Initializing none GPU layer.\n");
+  *num_devices = 1;
+}
+
+static bool chpl_gpu_device_alloc = false;
+
+void chpl_gpu_impl_on_std_modules_finished_initializing(void) {
+  // The standard module has some memory that we allocate when we  are "on" a
+  // GPU sublocale when in fact we want to allocate it on the device. (As of
+  // the writing of this comment this is in `helpSetupLocaleGPU` in
+  // `LocaleModelHelpSetup`).
+  //
+  // Basically during the setup of the locale model we need to be "on" a given
+  // sublocale when we instantiate the object for it (the expectation is that
+  // the wide pointer for a sublocale appears to be on that sublocale),
+  // but in practice we don't actually want the data for the GPU sublocale
+  // object to be on the GPU).
+  //
+  // It's a bit of a hack but to handle this we start off setting
+  // `chpl_gpu_device_alloc` to false indicating that we shouldn't actually
+  // do any allocations on the device. Once the standard modules have finished
+  // loading this callback function
+  // (`chpl_gpu_impl_on_std_modules_finished_initializing`) gets called and we
+  // flip the flag.
+  chpl_gpu_device_alloc = true;
+}
+
+bool chpl_gpu_impl_is_device_ptr(const void* ptr) {
+  return false;  // this OK? maybe we want assertions to go through?
+}
+
+bool chpl_gpu_impl_is_host_ptr(const void* ptr) {
+  return true;
+}
+
+inline void chpl_gpu_impl_launch_kernel(int ln, int32_t fn,
+                                        const char* name,
+                                        int grd_dim_x,
+                                        int grd_dim_y,
+                                        int grd_dim_z,
+                                        int blk_dim_x,
+                                        int blk_dim_y,
+                                        int blk_dim_z,
+                                        int nargs, va_list args) {
+}
+
+inline void chpl_gpu_impl_launch_kernel_flat(int ln, int32_t fn,
+                                             const char* name,
+                                             int num_threads,
+                                             int blk_dim,
+                                             int nargs,
+                                             va_list args) {
+}
+
+void* chpl_gpu_impl_memmove(void* dst, const void* src, size_t n) {
+  return chpl_memmove(dst, src, n);
+}
+
+void* chpl_gpu_impl_memset(void* addr, const uint8_t val, size_t n) {
+  return memset(addr, val, n);
+}
+
+void chpl_gpu_impl_copy_device_to_host(void* dst, const void* src, size_t n) {
+  chpl_memcpy(dst, src, n);
+}
+
+void chpl_gpu_impl_copy_host_to_device(void* dst, const void* src, size_t n) {
+  chpl_memcpy(dst, src, n);
+}
+
+void chpl_gpu_impl_copy_device_to_device(void* dst, const void* src, size_t n) {
+  chpl_memcpy(dst, src, n);
+}
+
+void* chpl_gpu_impl_comm_async(void *dst, void *src, size_t n) {
+  chpl_memcpy(dst, src, n);
```

Comment:
How liberal do we want to be with our warnings? We could have it warn here that with `CHPL_GPU=cpu` asynchronous communication will be made synchronous.

If nothing else document this in the tech note.

Reply:
I chose to document in the technote as a warning should be used in cases where the application behavior can change in a way that can generate incorrect results etc. Turning async memcpy into a blocking one is more innocent than that to me.